# Elastic constants

We compute *clamped-ion* elastic constants of a crystal using
the algorithmic differentiation density-functional perturbation theory (AD-DFPT) approach
as introduced in [^SPH25].

[^SPH25]:
    Schmitz, N. F., Ploumhans, B., & Herbst, M. F. (2025)
    *Algorithmic differentiation for plane-wave DFT: materials design, error control and learning model parameters.*
    [arXiv:2509.07785](https://arxiv.org/abs/2509.07785)

We consider a crystal in its equilibrium configuration, where all atomic forces
and stresses vanish.  Homogeneous strains $η$ are then applied
relative to this relaxed structure.
The elastic constants are derived from the stress-strain relationship.
In [Voigt notation](https://en.wikipedia.org/wiki/Voigt_notation),
the stress $\sigma$ and strain $\eta$ tensors are represented as 6-component vectors.
The elastic constants $C$ are then given by
the Jacobian of the stress with respect to strain, forming a $6 \times 6$ matrix
$$
  C = \frac{\partial \sigma}{\partial \eta}.
$$

The sparsity pattern of the matrix $C$ follows from crystal symmetry
and is tabulated in standard references (eg. Table 9 in [^Nye1985]).
This sparsity can be used a priori to reduce the number of strain patterns
that need to be probed to extract all independent components of $C$.
For example, cubic crystals have only three independent elastic constants
$C_{11}$, $C_{12}$ and $C_{44}$, with the pattern
$$
C = \begin{pmatrix}
  C_{11} & C_{12} & C_{12} & 0      & 0      & 0 \\
  C_{12} & C_{11} & C_{12} & 0      & 0      & 0 \\
  C_{12} & C_{12} & C_{11} & 0      & 0      & 0 \\
  0      & 0      & 0      & C_{44} & 0      & 0 \\
  0      & 0      & 0      & 0      & C_{44} & 0 \\
  0      & 0      & 0      & 0      & 0      & C_{44} \\
\end{pmatrix}.
$$
Thus we can just choose a suitable strain pattern $\dot{\eta} = (1, 0, 0, 1, 0, 0)^\top$,
such that $C\dot{\eta} = (C_{11}, C_{12}, C_{12}, C_{44}, 0, 0)^\top$. That is,
for cubic crystals like diamond silicon we obtain all independent elastic
constants from a single Jacobian-vector product on the stress-strain function.

[^Nye1985]:
     Nye, J. F. (1985).
     *Physical Properties of Crystals*. Oxford University Press.
     Comment: Since the elastic tensor transforms equivariantly under rotations,
     its numerical components depend on the chosen Cartesian coordinate frame.
     These tabulated patterns assume a standardized orientation of the structure
     with respect to conventional crystallographic axes.

This example computes the *clamped-ion* elastic tensor, keeping internal
atomic positions fixed under strain.  The *relaxed-ion* tensor includes
additional corrections from internal relaxations, which can be obtained
from first-order atomic displacements in DFPT (see [^Wu2005]).

[^Wu2005]:
    Wu, X., Vanderbilt, D., & Hamann, D. R. (2005).
    *Systematic treatment of displacements, strains, and electric fields in density-functional perturbation theory.*
    [Physical Review B, 72(3), 035105](https://doi.org/10.1103/PhysRevB.72.035105).

In [1]:
using DFTK
using PseudoPotentialData
using LinearAlgebra
using ForwardDiff
using DifferentiationInterface
using AtomsBuilder
using Unitful
using UnitfulAtomic


pseudopotentials = PseudoFamily("dojo.nc.sr.pbe.v0_4_1.standard.upf")
a0_pbe = 10.33u"bohr"  # Equilibrium lattice constant of silicon with PBE
model0 = model_DFT(bulk(:Si; a=a0_pbe); pseudopotentials, functionals=PBE())

Ecut = recommended_cutoff(model0).Ecut
kgrid = [4, 4, 4]
tol = 1e-6

function symmetries_from_strain(model0, voigt_strain)
    lattice = DFTK.voigt_strain_to_full(voigt_strain) * model0.lattice
    model = Model(model0; lattice, symmetries=true)
    model.symmetries
end

strain_pattern = [1., 0., 0., 1., 0., 0.];  # recovers [c11, c12, c12, c44, 0, 0]

For elastic constants beyond the bulk modulus, symmetry-breaking strains
are required. That is, the symmetry group of the crystal is reduced.
Here we simply precompute the relevant subgroup by applying the automatic
symmetry detection (spglib) to the finitely perturbed crystal.

In [2]:
symmetries_strain = symmetries_from_strain(model0, 0.01 * strain_pattern)


function stress_from_strain(model0, voigt_strain; symmetries, Ecut, kgrid, tol)
    lattice = DFTK.voigt_strain_to_full(voigt_strain) * model0.lattice
    model = Model(model0; lattice, symmetries)
    basis = PlaneWaveBasis(model; Ecut, kgrid)
    scfres = self_consistent_field(basis; tol)
    DFTK.full_stress_to_voigt(compute_stresses_cart(scfres))
end

stress_fn(voigt_strain) = stress_from_strain(model0, voigt_strain;
                                             symmetries=symmetries_strain,
                                             Ecut, kgrid, tol)
stress, (dstress,) = value_and_pushforward(stress_fn, AutoForwardDiff(),
                                           zeros(6), (strain_pattern,));

n     Energy            log10(ΔE)   log10(Δρ)   Diag   Δtime 
---   ---------------   ---------   ---------   ----   ------
  1   -8.453499543832                   -0.94    5.3    377ms
  2   -8.455628901749       -2.67       -1.77    1.0    171ms
  3   -8.455779267522       -3.82       -2.88    1.9    200ms
  4   -8.455790085540       -4.97       -3.36    2.9    277ms
  5   -8.455790180558       -7.02       -4.02    1.4    183ms
  6   -8.455790187255       -8.17       -4.99    1.8    203ms
  7   -8.455790187707       -9.34       -5.73    2.7    263ms
  8   -8.455790187713      -11.21       -6.18    2.1    225ms
Solving response problem
Iter  Restart  Krydim  log10(res)  avg(CG)  Δtime   Comment
----  -------  ------  ----------  -------  ------  ---------------
                                      64.2   5.04s  Non-interacting
   1        0       1        0.11     47.9   12.9s  
   2        0       2       -0.70     43.5   932ms  
   3        0       3       -1.71     37.4   533ms  


We can inspect the stress to verify it is small (close to equilibrium):

In [3]:
stress

6-element StaticArraysCore.SVector{6, Float64} with indices SOneTo(6):
 -2.312062775219399e-5
 -2.3120750684792833e-5
 -2.3120750684792806e-5
  4.793571035662018e-10
  0.0
  0.0

The response of the stress to `strain_pattern` contains the elastic constants
in atomic units, with the expected pattern $(c11, c12, c12, c44, 0, 0)$:

In [4]:
dstress

6-element StaticArraysCore.SVector{6, Float64} with indices SOneTo(6):
 0.005319941205521527
 0.002024899552955785
 0.002024899552955785
 0.00335190794426984
 0.0
 0.0

Convert to GPa:

In [5]:
println("C11: ", uconvert(u"GPa", dstress[1] * u"hartree" / u"bohr"^3))
println("C12: ", uconvert(u"GPa", dstress[2] * u"hartree" / u"bohr"^3))
println("C44: ", uconvert(u"GPa", dstress[4] * u"hartree" / u"bohr"^3))

C11: 156.51807371280427 GPa
C12: 59.57460153161377 GPa
C44: 98.61653624202599 GPa


These results can be compared directly to finite differences of the stress_fn:

In [6]:
h = 1e-3
dstress_fd = (stress_fn(h * strain_pattern) - stress_fn(-h * strain_pattern)) / 2h
println("C11 (FD): ", uconvert(u"GPa", dstress_fd[1] * u"hartree" / u"bohr"^3))
println("C12 (FD): ", uconvert(u"GPa", dstress_fd[2] * u"hartree" / u"bohr"^3))
println("C44 (FD): ", uconvert(u"GPa", dstress_fd[4] * u"hartree" / u"bohr"^3))

n     Energy            log10(ΔE)   log10(Δρ)   Diag   Δtime 
---   ---------------   ---------   ---------   ----   ------
  1   -8.453518693492                   -0.94    5.4    1.03s
  2   -8.455639313502       -2.67       -1.77    1.0    178ms
  3   -8.455783176285       -3.84       -2.89    1.9    199ms
  4   -8.455795197377       -4.92       -3.30    2.8    275ms
  5   -8.455795303198       -6.98       -3.82    1.3    168ms
  6   -8.455795314445       -7.95       -4.89    1.4    185ms
  7   -8.455795315188       -9.13       -5.41    2.8    272ms
  8   -8.455795315197      -11.03       -6.00    1.6    193ms
n     Energy            log10(ΔE)   log10(Δρ)   Diag   Δtime 
---   ---------------   ---------   ---------   ----   ------
  1   -8.453449462004                   -0.94    5.4    433ms
  2   -8.455606801752       -2.67       -1.77    1.0    226ms
  3   -8.455768527488       -3.79       -2.88    2.0    878ms
  4   -8.455782137695       -4.87       -3.29    2.9    272ms
  5   -8

Here are AD-DFPT results from increasing discretization parameters:

| Ecut | kgrid         | c11    | c12   | c44    |
|------|---------------|-------:|------:|-------:|
| 18   | [4, 4, 4]     | 156.51 | 59.57 |  98.61 |
| 18   | [8, 8, 8]     | 153.53 | 56.90 | 100.07 |
| 24   | [8, 8, 8]     | 153.26 | 56.82 |  99.97 |
| 24   | [14, 14, 14]  | 153.03 | 56.71 | 100.09 |

For comparison, Materials Project for PBE *relaxed-ion* elastic constants of
silicon [mp-149](https://next-gen.materialsproject.org/materials/mp-149):
c11 = 153 GPa, c12 = 57 GPa, c44 = 74 GPa.
Note the discrepancy in c44, which is due to us not yet including
ionic relaxation in this example.